Propagate route metadata in ext_authz

ikimonogakari opened the pull request:
Commit Message:
Add the ability to ext_authz that collect spcified namespaces from route metadata, and propagate them to external auth service. #30252

Additional Description:
The instruction of what namespace to select from route metadata, and the field in CheckRequest where the metadata context from route is filled are totally separate from those metadata context from connection or request.

Risk Level: Low
Testing: Unit tests
Docs Changes: Proto description updated.
Release Notes: Updated changelog
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

htuch commented on api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:
```diff
@@ -112,6 +112,14 @@ message ExtAuthz {
 
+  // Specifies a list of route metadata namespaces whose values, if present, will be passed to the
+  // ext_authz service. :ref:`filter_metadata <envoy_v3_api_field_config.core.v3.Metadata.filter_metadata>` is passed as an opaque ``protobuf::Struct``.
+  repeated string route_metadata_context_namespaces = 20;
```
I'm actually a bit confused (as an API reviewer and someone with experience in ext auth) on how the series of transforms occurs here. I think this gets added to the attribute context in the ext_authz request? But the attribute request metadata refers to this as dynamic metadata? Is this the same as the stream-level dynamic metadata?

At the very least, we need better docs here.

CC @kyessenov

ikimonogakari replied:
As discussed in #30252, we should treat request/connection dynamic metadata differently from route metadata.

It is transmit through attribute context in ext_authz request, but in a different field, other than the original `metadata_context`.
https://github.com/envoyproxy/envoy/pull/30477/files#diff-d5e4f508f08a9d9ca1c77377d37a1a746c66959da5358aa5d17a49821fadae0a

For docs, do you mean we need to explicitly tell users these namespaces for route_metadata will be saved in a different field in the request?

kyessenov replied:
@htuch This is referring to "route" xDS metadata, not the dynamic metadata. It's indeed confusing to call all of it metadata, which is why we thought a separate field might be better.

htuch replied:
https://github.com/envoyproxy/envoy/issues/30252#issuecomment-1766918389 suggests a separate field in the check request as well, not just in config.

ikimonogakari replied:
Updated the description of ext_authz API to point to [`route_metadata_context` at `attribute_context.proto`](https://github.com/envoyproxy/envoy/pull/30477/files#diff-d5e4f508f08a9d9ca1c77377d37a1a746c66959da5358aa5d17a49821fadae0a)

The presubmit failure should be caused by the `:ref: 'envoy_v3_api_field_service.auth.v3.checkrequest'`. I have no clue what the correct path to messages in `attribute_context.proto` is. Please educate me about how this ref path is constructed.

htuch replied:
Perfect, thanks.

htuch commented:
/lgtm api

htuch commented:
@kyessenov could you take the implementation review? Thanks.

kyessenov commented on test/extensions/filters/http/ext_authz/ext_authz_test.cc:
```diff
@@ -1508,6 +1511,174 @@ TEST_F(HttpFilterTest, ConnectionMetadataContext) {
+  EXPECT_THAT(check_request.attributes().route_metadata_context().filter_metadata(),
+              Not(Contains(Key("request.has.data"))));
+
+  for (const auto& namesapce_from_request :
```
nit: namesapce -> namespace and below

ikimonogakari replied:
Fixed.

kyessenov reviewed (dismissed):
Implementation LGTM.

htuch commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -15,6 +18,46 @@ namespace Extensions {
+
+using MetadataProto = ::envoy::config::core::v3::Metadata;
+
+void fillMetadataContext(std::vector<const MetadataProto*> source_metadata,
```
Nit: `const std::vector &`

ikimonogakari replied:
Changed.

htuch reviewed (commented):
LGTM, thanks, just one nit.

ikimonogakari commented:
Hi @htuch or @kyessenov, would any one of you mind giving another approval on this PR. I think the previous one expires as I rebased to the top and squashed commits, and it blocks merging this PR.

ggreenway commented:
> Hi @htuch or @kyessenov, would any one of you mind giving another approval on this PR. I think the previous one expires as I rebased to the top and squashed commits, and it blocks merging this PR.

In the future, please don't rebase commits (as documented https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md?plain=1#L136). Among other things, it dismisses approvals even if all you did was add additional changes from `main`.

htuch commented:
Yeah, force pushes are really bad, they break approvals and the review flow. We squash commits ourselves during merge.
